# Importing Libraries

In [1]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

# Preprocessing Data

In [2]:
flights = pd.read_csv('../datasets/final_flights.csv')
trains =  pd.read_csv('/Users/slavicagjorgieva/Desktop/TU/WS23/DOPP/Ex2/DOPP_group18/datasets/train_distances/combined_train_distances.csv')
trains

city1        city2  transit_distance [m]  transit_time [s]
0     San Bartolome    Barcelona                    -1                -1
1          Alicante    Barcelona                449965             28800
2         Amsterdam    Barcelona               1574426             55320
3             Spata    Barcelona               3359236            189858
4            Malaga    Barcelona               1049384             36900
...             ...          ...                   ...               ...
1464        Treviso        Rinas                982000             70058
1465            Ufa        Sochi               5760985            180660
1466            Ufa  Krasnoyarsk               4316816            162060
1467        Cologne        Varna               2360943            123578
1468         Zurich    Rovaniemi               3494234            156300

[1469 rows x 4 columns]

In [3]:
flights

route   time departure airport arrival airport  departure city  \
0     {'AAL', 'AMS'}   85.0               AAL             AMS         Aalborg   
1     {'AAQ', 'LED'}  170.0               AAQ             LED  Krasnyi Kurgan   
2     {'AMS', 'ABZ'}   85.0               AMS             ABZ       Amsterdam   
3     {'ABZ', 'LGW'}   95.0               ABZ             LGW        Aberdeen   
4     {'ACE', 'AGP'}  130.0               ACE             AGP   San Bartolomé   
...              ...    ...               ...             ...             ...   
1464  {'ZRH', 'OTP'}  150.0               ZRH             OTP          Zurich   
1465  {'ZRH', 'PRN'}  127.5               ZRH             PRN          Zurich   
1466  {'ZRH', 'RVN'}  205.0               ZRH             RVN          Zurich   
1467  {'ZRH', 'STR'}   45.0               ZRH             STR          Zurich   
1468  {'ZTH', 'ATH'}   62.5               ZTH             ATH       Zakynthos   

         arrival city           city1           city2  
0           Amsterdam         Aalborg       Amsterdam  
1      St. Petersburg  Krasnyi Kurgan  St. Petersburg  
2            Aberdeen       Amsterdam        Aberdeen  
3     Gatwick, Surrey        Aberdeen         Gatwick  
4              Málaga   San Bartolome          Malaga  
...               ...             ...             ...  
1464          Otopeni          Zurich         Otopeni  
1465        Prishtina          Zurich       Prishtina  
1466        Rovaniemi          Zurich       Rovaniemi  
1467        Stuttgart          Zurich       Stuttgart  
1468   Spata-Artemida       Zakynthos           Spata  

[1469 rows x 8 columns]

In [4]:
print("Number of rows in flights:", flights.shape[0])
print("Number of rows in trains:", trains.shape[0])


Number of rows in flights: 1469
Number of rows in trains: 1469


In [5]:
merged_data = pd.merge(flights, trains, on=['city1', 'city2'])
print("Number of rows in merged data:", merged_data.shape[0])

Number of rows in merged data: 1615


## Removing duplicates

In [6]:
cleaned_data = merged_data.drop_duplicates()
print("Number of rows after removing duplicates:", cleaned_data.shape[0])
cleaned_data.info()

Number of rows after removing duplicates: 1472
<class 'pandas.core.frame.DataFrame'>
Int64Index: 1472 entries, 0 to 1614
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   route                 1472 non-null   object 
 1   time                  1472 non-null   float64
 2   departure airport     1472 non-null   object 
 3   arrival airport       1472 non-null   object 
 4   departure city        1472 non-null   object 
 5   arrival city          1472 non-null   object 
 6   city1                 1472 non-null   object 
 7   city2                 1472 non-null   object 
 8   transit_distance [m]  1472 non-null   int64  
 9   transit_time [s]      1472 non-null   int64  
dtypes: float64(1), int64(2), object(7)
memory usage: 126.5+ KB


In [7]:
cleaned_data.describe()

time  transit_distance [m]  transit_time [s]
count  1472.000000          1.472000e+03       1472.000000
mean    127.859262          1.225550e+06      56415.970109
std      55.507938          1.104045e+06      48916.373427
min      10.000000         -1.000000e+00         -1.000000
25%      85.000000          2.333142e+05      13346.750000
50%     120.000000          1.045865e+06      48944.500000
75%     160.000000          1.910968e+06      83700.000000
max     402.500000          9.959629e+06     360180.000000

## Removing -1 (NaN) values for distance & time

In [8]:
filtered_data = cleaned_data[cleaned_data['transit_distance [m]'] != -1]
cleaned_df = filtered_data[filtered_data['transit_time [s]'] != -1]
cleaned_df
cleaned_df.to_csv('datasets/merged_df.csv')
cleaned_df

route        time departure airport arrival airport  \
0     {'AAL', 'AMS'}   85.000000               AAL             AMS   
2     {'AMS', 'ABZ'}   85.000000               AMS             ABZ   
3     {'ABZ', 'LGW'}   95.000000               ABZ             LGW   
23    {'AER', 'DME'}  215.000000               AER             DME   
27    {'AER', 'SVO'}  225.833333               AER             SVO   
...              ...         ...               ...             ...   
1608  {'ZRH', 'MUC'}   55.000000               ZRH             MUC   
1610  {'ZRH', 'OTP'}  150.000000               ZRH             OTP   
1611  {'ZRH', 'PRN'}  127.500000               ZRH             PRN   
1612  {'ZRH', 'RVN'}  205.000000               ZRH             RVN   
1613  {'ZRH', 'STR'}   45.000000               ZRH             STR   

     departure city     arrival city      city1      city2  \
0           Aalborg        Amsterdam    Aalborg  Amsterdam   
2         Amsterdam         Aberdeen  Amsterdam   Aberdeen   
3          Aberdeen  Gatwick, Surrey   Aberdeen    Gatwick   
23            Sochi           Moscow      Sochi     Moscow   
27            Sochi           Moscow      Sochi     Moscow   
...             ...              ...        ...        ...   
1608         Zurich           Munich     Zurich     Munich   
1610         Zurich          Otopeni     Zurich    Otopeni   
1611         Zurich        Prishtina     Zurich  Prishtina   
1612         Zurich        Rovaniemi     Zurich  Rovaniemi   
1613         Zurich        Stuttgart     Zurich  Stuttgart   

      transit_distance [m]  transit_time [s]  
0                  1027798             58200  
2                  1380434             65040  
3                   895395             39720  
23                 2919260            103440  
27                 2919260            103440  
...                    ...               ...  
1608                570891             29160  
1610               1869221            100620  
1611               1622133            114989  
1612               3494234            156300  
1613                454595             20040  

[1122 rows x 10 columns]

In [9]:
filtered_data = filtered_data.rename(columns={
    'time': 'flight_time', 
    'transit_time [s]': 'train_time', 
    'transit_distance [m]': 'train_distance'
})


In [10]:
# Convert train time from seconds to minutes using .loc
filtered_data.loc[:, 'train_time'] = filtered_data['train_time'] / 60

print(filtered_data.head())

             route  flight_time departure airport arrival airport  \
0   {'AAL', 'AMS'}    85.000000               AAL             AMS   
2   {'AMS', 'ABZ'}    85.000000               AMS             ABZ   
3   {'ABZ', 'LGW'}    95.000000               ABZ             LGW   
23  {'AER', 'DME'}   215.000000               AER             DME   
27  {'AER', 'SVO'}   225.833333               AER             SVO   

   departure city     arrival city      city1      city2  train_distance  \
0         Aalborg        Amsterdam    Aalborg  Amsterdam         1027798   
2       Amsterdam         Aberdeen  Amsterdam   Aberdeen         1380434   
3        Aberdeen  Gatwick, Surrey   Aberdeen    Gatwick          895395   
23          Sochi           Moscow      Sochi     Moscow         2919260   
27          Sochi           Moscow      Sochi     Moscow         2919260   

    train_time  
0        970.0  
2       1084.0  
3        662.0  
23      1724.0  
27      1724.0  


In [11]:
iata_codes_country = pd.read_csv('/Users/slavicagjorgieva/Desktop/TU/WS23/DOPP/Ex2/DOPP_group18/datasets/fixed-airports-codes.csv')

In [12]:
iata_codes_country

id    ident           type                       name  \
0        6523      00A       heliport          Total RF Heliport   
1      323361     00AA  small_airport       Aero B Ranch Airport   
2        6524     00AK  small_airport               Lowell Field   
3        6525     00AL  small_airport               Epps Airpark   
4      506791     00AN  small_airport       Katmai Lodge Airport   
...       ...      ...            ...                        ...   
78148   46378  ZZ-0001       heliport            Sealand Helipad   
78149  307326  ZZ-0002  small_airport  Glorioso Islands Airstrip   
78150  346788  ZZ-0003  small_airport      Fainting Goat Airport   
78151  342102     ZZZW         closed     Scandium City Heliport   
78152  313629     ZZZZ  small_airport     Satsuma Iōjima Airport   

       latitude_deg  longitude_deg  elevation_ft continent iso_country  \
0         40.070985     -74.933689          11.0       NaN          US   
1         38.704022    -101.473911        3435.0       NaN          US   
2         59.947733    -151.692524         450.0       NaN          US   
3         34.864799     -86.770302         820.0       NaN          US   
4         59.093287    -156.456699          80.0       NaN          US   
...             ...            ...           ...       ...         ...   
78148     51.894444       1.482500          40.0        EU          GB   
78149    -11.584278      47.296389          11.0        AF          TF   
78150     32.110587     -97.356312         690.0       NaN          US   
78151     69.355287    -138.939310           4.0       NaN          CA   
78152     30.784722     130.270556         338.0        AS          JP   

      iso_region         municipality scheduled_service gps_code iata_code  \
0          US-PA             Bensalem                no     K00A       NaN   
1          US-KS                Leoti                no     00AA       NaN   
2          US-AK         Anchor Point                no     00AK       NaN   
3          US-AL              Harvest                no     00AL       NaN   
4          US-AK          King Salmon                no     00AN       NaN   
...          ...                  ...               ...      ...       ...   
78148     GB-ENG              Sealand                no      NaN       NaN   
78149     TF-U-A     Grande Glorieuse                no      NaN       NaN   
78150      US-TX                 Blum                no     87TX       NaN   
78151      CA-YT  (Old) Scandium City                no      NaN       NaN   
78152      JP-46              Mishima                no     RJX7       NaN   

      local_code                                          home_link  \
0            00A  https://www.penndot.pa.gov/TravelInPA/airports...   
1           00AA                                                NaN   
2           00AK                                                NaN   
3           00AL                                                NaN   
4           00AN                                                NaN   
...          ...                                                ...   
78148        NaN                         http://www.sealandgov.org/   
78149        NaN                                                NaN   
78150       87TX                                                NaN   
78151        NaN                                                NaN   
78152       RJX7                                                NaN   

                                          wikipedia_link  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4                                                    NaN   
...                                                  ...   
78148  https://en.wikipedia.org/wiki/Principality_of_...   
78149                              

### Adding country code for each city (arrival & departure) and their geo-coordinates(longitude & latitude)

In [13]:
merged_departure = pd.merge(filtered_data, iata_codes_country, left_on='departure airport', right_on='iata_code', how='left')

# Rename columns from the first merge
merged_departure.rename(columns={'latitude_deg': 'dep_latitude_deg', 'longitude_deg': 'dep_longitude_deg', 'iso_country': 'dep_iso_country'}, inplace=True)

# Drop extra columns after the first merge
columns_to_drop_first_merge = ['id', 'ident', 'type', 'name', 'elevation_ft', 'continent', 'iso_region', 'municipality', 'scheduled_service', 'gps_code', 'iata_code', 'local_code', 'home_link', 'wikipedia_link', 'keywords']
merged_departure.drop(columns_to_drop_first_merge, axis=1, inplace=True)

# Second, merge for arrival airport
merged_data = pd.merge(merged_departure, iata_codes_country, left_on='arrival airport', right_on='iata_code', how='left')

# Rename columns from the second merge
merged_data.rename(columns={'latitude_deg': 'arr_latitude_deg', 'longitude_deg': 'arr_longitude_deg', 'iso_country': 'arr_iso_country'}, inplace=True)

# Drop extra columns after the second merge
columns_to_drop_second_merge = ['id', 'ident', 'type', 'name', 'elevation_ft', 'continent', 'iso_region', 'municipality', 'scheduled_service', 'gps_code', 'iata_code', 'local_code', 'home_link', 'wikipedia_link', 'keywords']
merged_data.drop(columns_to_drop_second_merge, axis=1, inplace=True)


print(merged_data.head())

            route  flight_time departure airport arrival airport  \
0  {'AAL', 'AMS'}    85.000000               AAL             AMS   
1  {'AMS', 'ABZ'}    85.000000               AMS             ABZ   
2  {'ABZ', 'LGW'}    95.000000               ABZ             LGW   
3  {'AER', 'DME'}   215.000000               AER             DME   
4  {'AER', 'SVO'}   225.833333               AER             SVO   

  departure city     arrival city      city1      city2  train_distance  \
0        Aalborg        Amsterdam    Aalborg  Amsterdam         1027798   
1      Amsterdam         Aberdeen  Amsterdam   Aberdeen         1380434   
2       Aberdeen  Gatwick, Surrey   Aberdeen    Gatwick          895395   
3          Sochi           Moscow      Sochi     Moscow         2919260   
4          Sochi           Moscow      Sochi     Moscow         2919260   

   train_time  dep_latitude_deg  dep_longitude_deg dep_iso_country  \
0       970.0         57.094763            9.84993              DK   


In [44]:
merged_data.describe()

flight_time  train_distance   train_time  dep_latitude_deg  \
count  1122.000000    1.122000e+03  1122.000000       1122.000000   
mean    121.597074    1.607852e+06  1233.580778         48.758990   
std      48.149078    9.920947e+05   714.170152          6.518959   
min      10.000000    5.296600e+04    79.000000         36.674900   
25%      85.000000    8.403885e+05   702.025000         43.472151   
50%     117.500000    1.500930e+06  1121.000000         49.476549   
75%     155.000000    2.156265e+06  1594.250000         52.308601   
max     325.000000    9.959629e+06  6003.000000         69.725800   

       dep_longitude_deg  arr_latitude_deg  arr_longitude_deg  \
count        1122.000000       1122.000000        1122.000000   
mean           12.108577         48.681017          11.378098   
std            13.381860          7.196403          12.785169   
min            -9.135920         36.151199          -9.355230   
25%             2.078460         43.192050           2.078460   
50%            11.100400         48.689899           9.709223   
75%            19.720600         52.453899          18.466200   
max            92.492437         70.065300          92.492437   

       train_time_hours  flight_time_plus_60  flight_time_hours_with_extra  
count       1122.000000          1122.000000                   1122.000000  
mean          20.559680           181.597074                      3.026618  
std           11.902836            48.149078                      0.802485  
min            1.316667            70.000000                      1.166667  
25%           11.700417           145.000000                      2.416667  
50%           18.683333           177.500000                      2.958333  
75%           26.570833           215.000000                      3.583333  
max          100.050000           385.000000                      6.416667

# Statistical Analysis

In [45]:
merged_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1122 entries, 0 to 1121
Data columns (total 19 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   route                         1122 non-null   object 
 1   flight_time                   1122 non-null   float64
 2   departure airport             1122 non-null   object 
 3   arrival airport               1122 non-null   object 
 4   departure city                1122 non-null   object 
 5   arrival city                  1122 non-null   object 
 6   city1                         1122 non-null   object 
 7   city2                         1122 non-null   object 
 8   train_distance                1122 non-null   int64  
 9   train_time                    1122 non-null   float64
 10  dep_latitude_deg              1122 non-null   float64
 11  dep_longitude_deg             1122 non-null   float64
 12  dep_iso_country               1122 non-null   object 
 13  arr

In [47]:
merged_data.isna()

route                           1122
flight_time                     1122
departure airport               1122
arrival airport                 1122
departure city                  1122
arrival city                    1122
city1                           1122
city2                           1122
train_distance                  1122
train_time                      1122
dep_latitude_deg                1122
dep_longitude_deg               1122
dep_iso_country                 1122
arr_latitude_deg                1122
arr_longitude_deg               1122
arr_iso_country                 1122
train_time_hours                1122
flight_time_plus_60             1122
flight_time_hours_with_extra    1122
dtype: int64

In [14]:
mean_flight_time = merged_data['flight_time'].mean()
mean_train_time = merged_data['train_time'].mean()

median_flight_time = merged_data['flight_time'].median()
median_train_time = merged_data['train_time'].median()

std_flight_time = merged_data['flight_time'].std()
std_train_time = merged_data['train_time'].std()


summary_data = {
    "Statistic": ["Mean", "Median", "Standard Deviation"],
    "Flight Time (hours)": [mean_flight_time / 60, median_flight_time / 60, std_flight_time / 60],
    "Flight Time (min)": [mean_flight_time, median_flight_time, std_flight_time],
    "Train Time (hours)": [mean_train_time / 60, median_train_time / 60, std_train_time / 60],
    "Train Time (min)": [mean_train_time, median_train_time, std_train_time]
}

summary_df = pd.DataFrame(summary_data)

print(summary_df)

            Statistic  Flight Time (hours)  Flight Time (min)  \
0                Mean             2.026618         121.597074   
1              Median             1.958333         117.500000   
2  Standard Deviation             0.802485          48.149078   

   Train Time (hours)  Train Time (min)  
0           20.559680       1233.580778  
1           18.683333       1121.000000  
2           11.902836        714.170152  


In [30]:
scatter_fig = px.scatter(merged_data, x='flight_time', y='train_time', hover_data=['city1', 'city2'])
scatter_fig.update_layout(title='Comparison of Flight Time vs Train Time in minutes', xaxis_title='Flight Time (minutes)', yaxis_title='Train Time (minutes)')
scatter_fig.show()

In [31]:
box_fig = go.Figure()
box_fig.add_trace(go.Box(y=merged_data['flight_time'], name='Flight Time'))
box_fig.add_trace(go.Box(y=merged_data['train_time'], name='Train Time'))
box_fig.update_layout(title='Distribution of Travel Times in minutes', yaxis_title='Time (minutes)')
box_fig.show()

In [19]:
shorter_train_routes = merged_data[merged_data['train_time'] < merged_data['flight_time']]

# Check if there are any such routes
if not shorter_train_routes.empty:
    print("Routes where train time is shorter than flight time:")
    print(shorter_train_routes[['city1', 'city2', 'flight_time', 'train_time']])
else:
    print("There are no routes where train time is shorter than flight time.")

There are no routes where train time is shorter than flight time.


In [26]:
merged_data['train_time_hours'] = merged_data['train_time'] / 60

# Add 60 minutes to flight time to account for pre-flight and post-flight activities
merged_data['flight_time_plus_60'] = merged_data['flight_time'] + 60

# Convert flight time (with the added 60 minutes) to hours
merged_data['flight_time_hours_with_extra'] = merged_data['flight_time_plus_60'] / 60

print(merged_data[['flight_time', 'train_time', 'train_time_hours', 'flight_time_plus_60', 'flight_time_hours_with_extra']].head())

   flight_time  train_time  train_time_hours  flight_time_plus_60  \
0    85.000000       970.0         16.166667           145.000000   
1    85.000000      1084.0         18.066667           145.000000   
2    95.000000       662.0         11.033333           155.000000   
3   215.000000      1724.0         28.733333           275.000000   
4   225.833333      1724.0         28.733333           285.833333   

   flight_time_hours_with_extra  
0                      2.416667  
1                      2.416667  
2                      2.583333  
3                      4.583333  
4                      4.763889  


In [37]:
scatter_fig = px.scatter(merged_data, x='flight_time_hours_with_extra', y='train_time_hours', hover_data=['city1', 'city2'])
scatter_fig.update_layout(title='Comparison of Flight Time vs Train Time in Hours',
                          xaxis_title='Flight Time (hours, including 60min extra)',
                          yaxis_title='Train Time (hours)')

# Define the maximum time value for plotting lines
max_time_hours = max(merged_data['flight_time_hours_with_extra'].max(), merged_data['train_time_hours'].max())

# Add line where flight time equals train time (y = x)
scatter_fig.add_shape(type='line',
                      line=dict(dash='dash', color='red'),
                      x0=0, y0=0, 
                      x1=max_time_hours, y1=max_time_hours)


# Show the plot
scatter_fig.show()

In [41]:
box_fig = go.Figure()
box_fig.add_trace(go.Box(y=merged_data['flight_time_hours_with_extra'], name='Flight Time'))
box_fig.add_trace(go.Box(y=merged_data['train_time_hours'], name='Train Time'))
box_fig.update_layout(title='Distribution of Travel Times in hours', yaxis_title='Time (hours)')
box_fig.show()

In [42]:
close_times = merged_data[(merged_data['flight_time_hours_with_extra'] - merged_data['train_time_hours']).abs() <= 1]

# Stack flight and train times into a single DataFrame with an identifier
stacked_times = pd.concat([
    close_times[['flight_time_hours_with_extra']].rename(columns={'flight_time_hours_with_extra': 'Time'}).assign(Type='Flight'),
    close_times[['train_time_hours']].rename(columns={'train_time_hours': 'Time'}).assign(Type='Train')
])

# Create the histogram using the stacked DataFrame
histogram_fig = px.histogram(stacked_times, x='Time', color='Type', barmode='overlay', nbins=50)

# Update the layout
histogram_fig.update_layout(
    title='Combined Histogram of Flight and Train Times (Hours)',
    xaxis_title='Time (hours)',
    yaxis_title='Count',
    legend_title_text='Type'
)

# Show the plot
histogram_fig.show()

In [28]:
shorter_train_routes = merged_data[merged_data['train_time'] < merged_data['flight_time_plus_60']]

# Check if there are any such routes
if not shorter_train_routes.empty:
    print("Routes where train time is shorter than flight time:")
    print(shorter_train_routes[['city1', 'city2', 'flight_time_plus_60', 'train_time']])
else:
    print("There are no routes where train time is shorter than flight time.")

Routes where train time is shorter than flight time:
                  city1              city2  flight_time_plus_60  train_time
292          Dusseldorf  Frankfurt am Main                110.0  109.000000
351                Rome            Firenze                115.0   85.000000
400   Frankfurt am Main         Dusseldorf                110.0  109.000000
420   Frankfurt am Main          Stuttgart                110.0   79.000000
734              Napoli               Rome                115.0   79.483333
1053           Valencia             Madrid                122.5  113.000000


notes for High Speed trains

Rome – Naples	-> max speed: 300 km/h(186 mph)

Florence – Rome -> max speed:	250 km/h (155 mph)

Cologne – Frankfurt -> max speed: 300 km/h (186 mph)

Valencia - Madrid -> max speed: 300 km/h (186 mph)

In [17]:
merged_data.to_csv('datasets/merged_df.csv')